spatialos-sdk/src/worker/schema/component_update.rs

```diff
@@ -0,0 +1,159 @@
+use crate::worker::{
+    component::{Component, ComponentId, Update},
+    schema::{owned::*, ArrayField, FieldId, Object, SchemaField},
+};
+use spatialos_sdk_sys::worker::*;
+use std::marker::PhantomData;
+
+#[derive(Debug)]
+pub struct ComponentUpdate(PhantomData<*mut Schema_ComponentUpdate>);
+
+impl ComponentUpdate {
+    pub fn new<U: Update>(update: &U) -> Owned<Self> {
+        // Create the underlying `Schema_ComponentUpdate` and wrap it in a smart pointer.
+        let mut result: Owned<Self> =
+            unsafe { Owned::new(Schema_CreateComponentUpdate(U::Component::ID)) };
+
+        // Populate the update object from the update data.
+        let component_update = &mut *result;
+        update.into_update(component_update);
+
+        result
+    }
+
+    pub(crate) unsafe fn from_raw<'a>(raw: *mut Schema_ComponentUpdate) -> &'a Self {
+        &*(raw as *mut _)
+    }
+
+    pub fn component_id(&self) -> ComponentId {
+        unsafe { Schema_GetComponentUpdateComponentId(self.as_ptr()) }
+    }
+
+    pub fn deserialize<U: Update>(&self) -> U {
+        // TODO: Does it make sense to assert here? Or should we allow for runtime recovery
```

Comment:
Depends who calls this - if its library code exclusively, then I'd say assert as it indicates a critical flaw in the library.

If this is a user-facing API, then I'd say return a `Result<U, SomeError>`

Reply:
Ah, that's a good point! I'll have a look at how this function is used and whether or not it needs to be user-facing. I think right now it's only used internally within `ComponentUpdateOp`, but if there's a reasonable use case for users to want to call it, then I'll have it return a `Result` 😁

Reply:
I opted to return an `Option<U>` since that's what I'm doing in similar scenarios 👍